src/GitHub.VisualStudio/AssemblyResolverPackage.cs

```diff
@@ -1,90 +1,124 @@
 ﻿using System;
 using System.IO;
-using System.Linq;
 using System.Reflection;
 using System.Globalization;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
 using System.Runtime.InteropServices;
 using Microsoft.VisualStudio;
 using Microsoft.VisualStudio.Shell;
+using NLog;
 
 namespace GitHub.VisualStudio
 {
-    // This is the Git service GUID, which fires early and is used by GitHubService.
-    //[ProvideAutoLoad(Guids.GitSccProviderId)]
     // This fires before ShellInitialized and SolutionExists.
-    //[ProvideAutoLoad(VSConstants.UICONTEXT.NoSolution_string)]
+    [ProvideAutoLoad(VSConstants.UICONTEXT.NoSolution_string)]
     [Guid(Guids.guidAssemblyResolverPkgString)]
     public class AssemblyResolverPackage : Package
     {
-        // list of assemblies that should be resolved by name only
-        static readonly string[] ourAssemblies =
-        {
-            // resolver is required for these
-            "GitHub.UI",
-            "GitHub.VisualStudio.UI",
-
-            // these are signed by StrongNameSigner
-            "Markdig",
-            "Markdig.Wpf",
+        // list of assemblies that should be considered when resolving
+        IEnumerable<ProvideDependentAssemblyAttribute> dependentAssemblies;
+        string packageFolder;
 
-            // these are included just in case
-            "GitHub.UI.Reactive",
-            "System.Windows.Interactivity"
-        };
-
-        readonly string extensionDir;
+        IDictionary<string, Assembly> resolvingAssemblies;
+        IDictionary<string, Exception> resolvingExceptions;
 
         public AssemblyResolverPackage()
         {
-            extensionDir = Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location);
-            AppDomain.CurrentDomain.AssemblyResolve += LoadAssemblyFromExtensionDir;
+            var asm = Assembly.GetExecutingAssembly();
+            packageFolder = Path.GetDirectoryName(asm.Location);
+            dependentAssemblies = asm.GetCustomAttributes<ProvideDependentAssemblyAttribute>();
+            resolvingAssemblies = new Dictionary<string, Assembly>();
+            resolvingExceptions = new Dictionary<string, Exception>();
+            AppDomain.CurrentDomain.AssemblyResolve += ResolveAssemblyFromPackageFolder;
         }
 
         protected override void Dispose(bool disposing)
         {
-            AppDomain.CurrentDomain.AssemblyResolve -= LoadAssemblyFromExtensionDir;
+            AppDomain.CurrentDomain.AssemblyResolve -= ResolveAssemblyFromPackageFolder;
+
+            if (resolvingAssemblies.Count > 0 || resolvingExceptions.Count > 0)
+            {
+                // Avoid executing any logging code unless there is something to log.
+                WriteToLog();
```

Comment:
Why do we log the resolved assemblies and resolve errors when VS is shutdown rather than when they occur?

Reply:
For a few reasons:

1. Since this package is always loaded, I wanted to avoid loading any extra assemblies as VS starts.
2. Avoid the possability of recursive assembly resolving (in a previous version there was an issues with `Trace.WriteLine`).
3. Avoid doing any file creation or writes as VS starts.

I've tried to make it as predictable and light weight as possible. :-)